### PR TITLE
luks.nix | TPM implementation

### PR DESCRIPTION
**Pull request by Sk7Str1pe**
Implements support for TPM decryption.

### CODE REVIEW

**Comment by iFreilicht** [comment]
Instead of committing here and waiting for buildbot every time, you can also run tests locally, see https://github.com/nix-community/disko/blob/master/docs/testing.md

**Comment by Sk7Str1pe** [comment]
> Instead of committing here and waiting for buildbot every time, you can also run tests locally, see https://github.com/nix-community/disko/blob/master/docs/testing.md

got you, thanks. this way really annoying to wait buildbot. last commit should be built without any errors

**Comment by Sk7Str1pe** [comment]
yea i added odd function in test.nix and it was crashing test, dammit. PR polished and ready to be merged

**Comment by Sk7Str1pe** [comment]
btw @iFreilicht if TPM used, should there be an options to disable passphrase input, or the opposite, requires recovery password?

**Comment by iFreilicht** [comment]
> btw @iFreilicht if TPM used, should there be an options to disable passphrase input, or the opposite, requires recovery password?

I think if you disable passphrase input by default, it will be impossible to create the LUKS partition, right? You always have to first encrypt it with a passphrase or keyfile, and enrolling the TPM can only be done afterwards.

File: example/luks-btrfs-tpm.nix [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+                type = "luks";
+                name = "crypted";
+                # disable settings.keyFile if you want to use interactive password entry
+                #passwordFile = "/tmp/secret.key"; # Interactive
+                settings = {
+                  allowDiscards = true;
+                  keyFile = "/tmp/secret.key";
+                };
+                tpmEnroll = true;
+                additionalKeyFiles = [ "/tmp/additionalSecret.key" ];
```
[the remark this answers is not in the document]
Comment (Sk7Str1pe) [reply example/luks-btrfs-tpm.nix:32]: tpm is not enabled in qemu by default. sadly i have no idea how to enable it in test vm. but i managed to test this on my custom vm and it succeeded (tpm token successfully added to luks header) 
tell me how do i enable tpm in test build pls.

File: lib/types/luks.nix [hunk 2]
```diff
@@ -169,7 +195,7 @@ in
         };
     };
     _config = lib.mkOption {
-      internal = true;
+      internal = true;
```
[the remark this answers is not in the document]
Comment (Sk7Str1pe) [reply lib/types/luks.nix:198]: done.

File: lib/types/luks.nix [hunk 3]
```diff
@@ -133,6 +143,22 @@ in
...
+              check="$(ls /dev/tpm*)"
+              if [[ -n $(check) ]] ; then
+                echo 0;
+		return;
+	      fi
+	      echo 1;
+	      return;
+	    }
+	    if [[ $(TPMParse) == "0" ]] ;then
+	      ${sdCryptEnroll}
```
[the remark this answers is not in the document]
Comment (Sk7Str1pe) [reply lib/types/luks.nix]: done.

File: lib/types/luks.nix [hunk 4]
```diff
@@ -27,6 +27,11 @@ let
       ${toString config.extraOpenArgs} \
       ${keyFileArgs} \
   '';
+  sdCryptEnroll = ''
+    systemd-cryptenroll \
+    ${lib.optionalString (keyFile != null) "--unlock-key-file ${keyFile}"} \
```
[the remark this answers is not in the document]
Comment (Sk7Str1pe) [reply lib/types/luks.nix:32]: again, i couldnt test this with nix build but it succeeded on my custom vm

File: example/luks-btrfs-tpm.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (iFreilicht) [reply example/luks-btrfs-tpm.nix:32]: The option is called `virtualisation.tpm.enable`. 

You should probably add a `enableTPM` parameter [here](https://github.com/nix-community/disko/blob/master/lib/tests.nix#L57) and then set this option based on that value [here](https://github.com/nix-community/disko/blob/master/lib/tests.nix#L219). Make it false by default to keep boot times low on other tests.

**Comment by Sk7Str1pe** [comment]
Ahhh @iFreilicht I just realized there's one VERY IMPORTANT ISSUE with TPM enrollment. It logs every event, so enrolling token in installation medium is useless, TPM will deny opening luks due to different bootloader. Currently I have no idea how we can solve this.

**Comment by iFreilicht** [comment]
Hmm indeed, it seems the enrollment uses the TPM's PCR7 register (which is the Secure Boot chain, including all certificates every loader in the boot chain was signed with) as a key, so if any part of this changes, it will fail to unlock the luks partition, so luks falls back to passphrase/key file and you have to re-enroll the TPM.

To me it sounds like this is a problem that disko can't solve, tbh. Maybe nixos-anywhere could, if it's logging in again after the reboot (not sure if it does), but then that seems like a bit of magic that might be dangerous to hide from users.

So I'd say we close this PR and the issue you opened as "won't fix". I'll post this explanation there.

Thank you for putting in the time though, this is very valuable information to have!

**Pull request closed without merging**